scripts/speech_recognition/oomptimizer.py

```diff
@@ -0,0 +1,283 @@
+    In order to simplify finding the optimal settings, OOMptimizer scans each bucket to find the maximum possible
+    batch size that doesn't trigger a CUDA OOM.
+
+    \b
+    The suggested workflow is the following:
+    1) Run scripts/speech_recognition/estimate_duration_bins.py to get the duration distribution of your data.
+    2) Run OOMptimizer to find the optimal batch sizes for your specific model, optimizer, and GPU.
+    3) Use these optimal settings in your actual training script and enjoy optimal GPU utilization OOM-free.
+    """
+    logging.setLevel(logging.CRITICAL)
+    device = "cuda"
+
+    print("Intializing ASR model.")
+    # TODO(pzelasko): This currently only supports "from_pretrained".
+    #                 We need to be able to read a model training configuration and instantiate the model
+    #                 and figure out all the necessary details from that.
+    trainer = pl.Trainer(barebones=True)
+    trainer.log_every_n_steps = 1000000
+    model = ASRModel.from_pretrained(model_name, trainer=trainer).to(device)
+
+    # TODO(pzelasko): ideally move into model @property e.g. "oomptimizer_schema" :D
+    schema = [
+        {"type": NeuralType(("B", "T"), AudioSignal()), "seq_length": "input"},
+        {"type": NeuralType(("B",), LengthsType()), "seq_length": "input"},
+        {
+            "type": NeuralType(("B", "T"), LabelsType()),
+            "seq_length": "output",
+            "vocab_size": model.tokenizer.vocab_size,
+        },
+        {"type": NeuralType(("B",), LengthsType()), "seq_length": "output"},
+    ]
+    if isinstance(model, EncDecMultiTaskModel):
+        schema.extend(
+            [{"type": "dummy"}, {"type": "dummy"}]
+        )  # multi-task has 2 extra tensors not needed for batch size tuning
+
+    print("Setting up the optimizers.")
+    optimizer, _ = model.setup_optimization({"name": optimizer_name, "lr": 1e-7, "weight_decay": 0.0})
+
+    def get_max_seq_lens(buckets):
+        # TODO(pzelasko): support text data inputs.
+        return [
+            (
+                compute_num_samples(d, sampling_rate=16000),  # num_samples
+                math.ceil(labels_per_second * d),  # num_labels; might need to go data-driven for optimal tuning
+            )
+            for d in buckets
+        ]
+
+    print("Starting profiling.")
+    max_seq_lens = get_max_seq_lens(buckets)
+    gen = ProfilingBatchGenerator(schema=schema, start_batch_size=start_batch_size, rel_gap_thresh=threshold)
+    profile = {}
+
+    for seq_len_in, seq_len_out in max_seq_lens:
+        print(f"The current sequence lengths are: input={seq_len_in} output={seq_len_out}.")
+        gen.reset()
+        batch_idx = 0
+
+        def step():
+            batch = gen(seq_len_in, seq_len_out)
+            oom = False
+            try:
+                print(f"Current gap: {gen.current_rel_gap}. Attempting shapes: {[b.shape for b in batch]}", end=" ")
+                optimizer.zero_grad()
```

Comment:
It is theoretically possible to do these three lines in a cuda stream capture with "relaxed" mode to avoid doing any sort of GPU-side computation. However, it will work only for code that has no data-dependent shapes (like torch.nonzero). Note that I haven't run your code and don't know how slow it is right now.

Reply:
It is surprisingly fast - for ~30 buckets the total runtime seems within 1-2 minutes. If CUDA graph "relaxed" mode would be "ok" with skipping NCCL ops then we might even incorporate this as a training time calibration (which we can't do now because these steps trigger NCCL syncs, if one GPU dies and other doesn't, it would hang). But even as-is I think this is a viable approach.

Reply:
For lots of buckets (i.e. 100+) it takes a while. We should try the "relaxed" CUDA graph trick, and if it works, make a follow up PR.

Reply:
just curious how long "a while" is.

The relaxed cuda graph trick definitely won't always work unfortunately... I spoke with someone who works on end-to-end training and he told me that there is a cudaStreamSynchronize() is the torch.amp.GradScaler, which will prevent using relaxed stream capture for models that do gradient scaling in mixed precision training.

Reply:
I think it's around 15-20 minutes for 150 buckets.